File: java/common/src/main/java/common/StringUtils.java
package common;

/**
 * 
 * This holds string utilities. Operations include:
 * - padding a message to the left
 * - padding a message to the right
 * - padding a message to the center
 * 
 * - zero filling an integer (e.g. 12 (with size 3) becomes 012)
 * - a test driver to test that all the functions are working
 * 
 */

public class StringUtils {

    /**
     * 
     * Pads a message to the left
     * 
     * @param message
     * @param width the width of the message plus padding
     * @return a string with the message padded to the left
     */
    public static String padLeft(String message, int width) {
        assert width > message.length();

        String spaces = " ".repeat(width - message.length());

        return spaces + message;
    }

    /**
     * 
     * Pads a message to the right
     * 
     * @param message
     * @param width the width of the message plus padding
     * @return a string with the message padded to the right
     */
    public static String padRight(String message, int width) {
        assert width > message.length();

        String spaces = " ".repeat(width - message.length());

        return message + spaces;
    }

    /**
     * 
     * Pads a message to the center
     * 
     * @apiNote if the width ends with a odd number (after the message), the extra
     *          space not be accounted for
     * 
     * @param message
     * @param width the width of the message plus padding
     * @return a string with the message padded to the center
     */
    public static String padCenter(String message, int width) {
        assert width > message.length();

        String spaces = " ".repeat((width - message.length()) / 2);

        return spaces + message + spaces;
    }

    /**
     * 
     * Zero fills a message with size many zeros
     * 
     * @param in
     * @param width
     * @return the message zero filled (size many zeros)
     */
    public static String zeroFill(String in, int width) {
        if (in.length() >= width)
            return in;

        return "0".repeat(width - in.length()) + in;
    }

    public static void stringUtilsTestDriver() {
        System.out.println("String utils test");

        // 14 characters is "something cool"
        String msgLeftPadded = padLeft("something cool", 18);
        assert msgLeftPadded.equals("    something cool") : "Message was not padded correctly";
        System.out.printf("Message padded left: %s\n", msgLeftPadded);

        String msgRightPadded = padRight("something cool", 20);
        assert msgRightPadded.equals("something cool      ") : "Message was not padded correctly";
        System.out.printf("Message padded right: %s\n", msgRightPadded);

        String msgCenterPadded = padCenter("something cool", 21);
        assert msgCenterPadded.equals("   something cool   ") : "Message was not padded correctly";
        System.out.printf("Message padded center: %s\n", msgCenterPadded);

        String zeroFillStr = zeroFill("189", 4);
        System.out.printf("Zero filled (str): %s\n", zeroFillStr);
        assert zeroFillStr.equals("0189") : "Zero fill did not return a correct zero fill string";

        System.out.println("String utils test success");
    }

}

/**
 * 
 * Sample Output
 * 
 * String utils test
 * Message padded left:     something cool
 * Message padded right: something cool      
 * Message padded center:    something cool   
 * Zero filled (str): 0189
 * String utils test success
 * 
 */

File: cpp/common/src/StringUtils.cpp
#include "../include/common/StringUtils.hpp"

#include <iostream>
#include <cmath>
#include <cassert>

namespace common {

    // pads the message to the left (e.g. "something" becomes "   something", based off the width)
    std::string padLeft(const std::string& message, int width) {
        assert(width > message.length());
        
        std::string s(width - message.length(), ' ');
        return s + message;
    }

    // pads the message to the right (e.g. "something" becomes "something   ", based off the width)
    std::string padRight(const std::string& message, int width) {
        assert(width > message.length());

        std::string s(width - message.length(), ' ');
        return message + s;
    }

    // pads the message to the center (e.g. "something" becomes "   something   ", based off the width)
    std::string padCenter(const std::string& message, int width) {
        assert(width > message.length());
        
        std::string s((width - message.length()) / 2, ' ');
        return s + message + s;
    }

    // fills a string with zeros (e.g. 12, (size 3) becomes (012))
    std::string zeroFill(int input, size_t size) {
        std::string in = std::to_string(input);
        
        if (in.length() >= size)
            return in;

        return std::string(size - in.length(), '0') + in;
    }

    // tests that the string utils works correctly
    void stringUtilsTestDriver() {
        std::cout << "String utils test \n";

        std::string msgLeftPadded = padLeft("something cool", 18);
        assert(msgLeftPadded == "    something cool");
        std::cout << "Message padded left: >" << msgLeftPadded << "<\n";

        std::string msgRightPadded = padRight("something cool", 20);
        assert(msgRightPadded == "something cool      ");
        std::cout << "Message padded right: >" << msgRightPadded << "<\n";

        std::string msgCenterPadded = padCenter("something cool", 21);
        assert(msgCenterPadded == "   something cool   ");
        std::cout << "Message padded center: >" << msgCenterPadded << "<\n";

        std::string zeroFillStr = zeroFill(123, 5);
        std::cout << "Zero filled (str): " << zeroFillStr << "\n";
        assert(zeroFillStr == "00123");

        std::cout << "String utils test success\n";
    }

} // namespace common
